Try reconnect if connection aborted or cannot be established

johahauf opened the pull request.

samhed commented:
Continued from #1186

samhed reviewed (changes requested):
I've tested it now and it works great as far as I can tell. However, it seems like you have included a lot of irrelevant and unwanted changes in this commit? Could you look that over?

johahauf commented:
Hi @samhed,
thanks for testing and the positive feedback. However I am not sure which changes you think are unrelated? I reviewed all the error messages and updated the comments where necessary. Also I changed the error handling from va_args to a std::string as this makes the code a way simpler an eases the handling of the error messages.

samhed commented:
> Also I changed the error handling from va_args to a std::string as this makes the code a way simpler an eases the handling of the error messages.

That's what I think is unrelated, the code uses `va_args` in many places, if you want this sort of change that fits into it's own PR, not here. Also, you can't just remove the call to the translation function `_()`.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -87,9 +87,8 @@ char vncServerName[VNCSERVERNAMELEN] = { '\0' };
-static bool inMainloop = false;
 static bool exitMainloop = false;
-static const char *exitError = NULL;
+string exitError;
```
Should still be static.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -113,24 +112,19 @@ void exit_vncviewer(const char *error, ...)
-      fl_alert("%s", exitError);
-    exit(EXIT_FAILURE);
-  }
+  exitError = errorString;
```
I don't think it's legal to assign NULL to a string.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -633,12 +651,12 @@ int main(int argc, char** argv)
 
   if (listenMode) {
-    std::list<SocketListener*> listeners;
+    list<SocketListener*> listeners;
```
Please avoid unrelated cleanups like this in the same commit.

johahauf commented on vncviewer/CConn.cxx:
```diff
@@ -272,12 +272,16 @@ void CConn::socketEvent(FL_SOCKET fd, void *data)
+      exit_vncviewer(_("Connection error: %s\n\n"), e.str());
     }
+    recursing = false;
+    return;
```
exit_vncviewer is no longer calling exit immediately so recursing and return is needed here.

CendioOssman replied:
`exit_vncviewer()` only calls `exit()` during early errors. In a case like this it would just break the mainloop.

johahauf replied:
actually `exit_vncviewer()` in this version never calls `exit` but only breaks the mainloop. The mainloop than waits in the re-connect dialog for the user. During this wait for the user input I saw the error message again and again...

CendioOssman replied:
Hmm... that suggests some cleanup isn't done. This should get cleared by the destructor for `CConn` once it goes out of scope.

I think you'll have to debug a bit more to see why this error message keeps repeating.

CendioOssman commented:
Have you been able to have a look at the comments, @johahauf?

johahauf commented:
Sorry, was busy the last weeks, but updated the solution now to differentiate between exit due to interrupt and abort.
Please have a look and let me know your opinion.

samhed reviewed (approved):
Looks good, seems you have addressed all @CendioOssman's comments.

I've tested the functionality again and it works as advertised. I tested a connection to a VNC server running on a VM and disconnecting the machine from the network. After a while vncviewer gave up and displayed the alert. After enabling the network again I chose to reconnect which worked well.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -109,28 +108,47 @@ static const char *about_text()
   }
 
-  if (inMainloop)
-    exitMainloop = true;
```
This function could unfortunately be called when the mainloop has been started, so this still needs to be handled.

CendioOssman replied:
That's my point, it shouldn't be calling `exit()` and just abruptly terminating things. At least not if we've gotten to the point of the mainloop. Getting the ability to clean things up is always preferred.

johahauf replied:
`exit_vncviewer` is called before the main loop is entered in vncviewer.cxx. Only in touch.cxx it might be called when we are already in the main loop and I read the code like it should have all cases a terminating character (`return; //Not reached` is directly after the call to `exit_vncviewer`). So I see no need for a graceful shutdown. While writing this maybe the better name is to call this `terminate_vncviewer` to make it clear that this is not a graceful shutdown.

CendioOssman replied:
That comment was probably just a copy-paste error.

Shutting down as gracefully as we can should always be preferred unless there is some technical limitations preventing us from doing so. Otherwise we might leave stray stuff around, not signal other programs properly, or various other issues.

johahauf replied:
Than my proposal would be to limit `exit_vncviewer` to vncviewer.cxx and have all other places to call `abort_vncviewer` which does a graceful termination/reconnect. Is this the direction you are thinking?

CendioOssman replied:
No, I still think it's a good separation of one aborting the connection, and the other aborting the viewer. There are some errors where a reconnect will not be able to fix things (hopefully few though).

So this function here should be able to break the mainloop in a way that does not trigger a reconnect. A simple extra flag should be able to solve that.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -109,28 +108,47 @@ static const char *about_text()
+void exit_vncviewer(const char *error, ...)
+{
+  // Prioritise the first error we get as that is probably the most
+  // relevant one.
```
Same thing here. This comment is no longer respected.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -144,20 +162,43 @@ void about_vncviewer()
   }
 }
 
+static void mainloop(const char* vncserver, network::Socket* sock)
```
`mainloop()` and `run_mainloop()` are a bit too similar. Can we tweak the naming to clarify what the difference is?

johahauf replied:
No good idea.... I tend to merge `run_mainloop` and `mainloop` into one method, as `run_mainloop` is just about 10 lines of code and `mainloop` is also small. Would you agree?

CendioOssman replied:
Works for me.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -678,9 +717,8 @@ int main(int argc, char** argv)
           }
       }
     } catch (rdr::Exception& e) {
-      vlog.error("%s", e.str());
```
Did you have a specific plan when moving these?

johahauf replied:
abort_vncviewer is already logging the error so no need to do it twice I thought.

CendioOssman replied:
The message given to `abort_vncviewer()` is something we show to the user, which isn't always the same thing we want to put in a log. So I think it is best to separate the two, just like we had before.

johahauf replied:
OK fine, it just looked inconsistent to me that sometimes something is logged and in other cases not and thought the easiest is to have the logging at a central place.

CendioOssman commented on vncviewer/vncviewer.cxx:
```diff
@@ -109,28 +108,47 @@ static const char *about_text()
+}
+
+
+void exit_vncviewer(const char *error, ...)
```
This should probably be renamed given that it no longer reliably exits vncviewer.

CendioOssman replied:
Referring to "vncviewer" might no longer be relevant for this function as it is now more about terminating the current connection, not the entire viewer.

CendioOssman reviewed (changes requested):
Sorry for the delay. I missed that you had updated the code.

There are still a few tweaks I'd like to see. But I think all the main parts look good now. So some last polish and we can merge this. :)

CendioOssman commented:
Just a heads up that I'll need some time before I can have a look at this as I'm currently swamped with Cendio stuff.

CendioOssman commented:
I feel like we're going in circles a bit here, so I've adjusted the things I feel need changing and committed everything as 44b085adee9a0ecf697b48e38d2434d920385928. So this is effectively merged now.
